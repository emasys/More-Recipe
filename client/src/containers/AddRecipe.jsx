import React, { Component } from 'react';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { addRecipe } from '../actions';
import Dropzone from 'react-dropzone';
import { CloudinaryContext, Transformation, Image } from 'cloudinary-react';
import axios from 'axios';
import config from '../config';

//component
import Navbar from '../components/Navbar';

class AddRecipe extends Component {
  constructor(props) {
    super(props);

    this.state = { preview: null, files: null, fileURL: null, status: 'fade' };
    this.handleForm = this.handleForm.bind(this);
    this.handleDrop = this.handleDrop.bind(this);
    this.handleDropRejected = this.handleDropRejected.bind(this);
    // this.uploadWidget = this.uploadWidget.bind(this);
    this.sendData = this.sendData.bind(this);
  }

  sendData(e) {}
  handleForm(e) {
    e.preventDefault();
    let data = {
      name: e.target.elements.recipe.value.trim(),
      ingredients: e.target.elements.ingredients.value.trim(),
      direction: e.target.elements.direction.value.trim(),
      description: e.target.elements.description.value.trim(),
      category: e.target.elements.category.value,
    };
    this.setState({
      status: 'show',
    });
    const { files } = this.state;
    // Push all the axios request promise into a single array
    const uploaders = files.map(file => {
      // Initial FormData
      const formData = new FormData();
      formData.append('file', file);
      formData.append('tags', `morerecipe`);
      formData.append('upload_preset', config.UPLOAD_PRESET);
      formData.append('api_key', config.API_KEY);
      formData.append('timestamp', (Date.now() / 1000) | 0);

      return axios
        .post('https://api.cloudinary.com/v1_1/emasys/image/upload', formData, {
          headers: { 'X-Requested-With': 'XMLHttpRequest' },
        })
        .then(response => {
          const resdata = response.data;
          data.foodImg = resdata.secure_url;
        });
    });

    axios.all(uploaders).then(() => {
      // ... perform after upload is successful operation

      // console.log('upload complete');
      this.props.addRecipe(data).then(() => {
        this.componentDidUpdate();
      })
    });
  }
  componentDidUpdate() {
    if (this.props.new_recipe.new_recipe) {
      if (this.props.new_recipe.new_recipe.recipe) {
        const id = this.props.new_recipe.new_recipe.recipe.id;
        this.props.history.push(`/recipe/${id}`);
      }
    }
  }
  handleDrop(files) {
    const [{ preview }] = files;
    // console.log(files);
    this.setState({ preview, files });
  }
  handleDropRejected(...args) {
    return console.log('reject', args);
  }

  render() {
    const { preview, status } = this.state;
    const recipeCategory = [
      'Breakfast',
      'Brunch',
      'Lunch',
      'Snacks',
      'Appetisers',
      'Dinner',
      'Soups',
      'Noodles',
      'Rice',
      'Pasta',
      'Meat',
      'Poultry',
      'Seafood',
      'Vegetarian',
      'Salads',
      'Sides',
      'Sauces',
      'Baking',
      'Desserts',
      'Drinks',
    ];
    // console.log(recipeCategory);

    return (
      <section className="container ">
        <Navbar />
        <div className="row justify-content-center mt-80">
          <div className="catalog-wrapper p-15">
            <form onSubmit={this.handleForm}>
              <ul className="form row">
                <li className="col-lg-6 col-sm-12">
                  <label>Recipe Name</label>
                  <input
                    type="text"
                    name="recipe"
                    required
                    className="col-lg-11 col-sm-12"
                    id="inputRecipe"
                    placeholder="Name of recipe"
                  />
                </li>
                <li className="col-lg-6 col-sm-12">
                  <label htmlFor="ingredients" className="col-form-label">
                    Ingredients
                  </label>
                  <textarea
                    placeholder="Add your ingredients and separate with a comma ','"
                    className="col-lg-11 col-sm-12"
                    id="ingredients"
                    rows="4"
                    required
                    name="ingredients"
                  />
                </li>
                <li className="col-lg-6 col-sm-12">
                  <label htmlFor="direction">Direction</label>
                  <textarea
                    className="col-lg-11 col-sm-12"
                    placeholder="how to make it happen"
                    id="direction"
                    rows="4"
                    required
                    name="direction"
                  />
                </li>
                <li className="col-lg-6 col-sm-12">
                  <label htmlFor="description">Description</label>
                  <textarea
                    className="col-lg-11 col-sm-12"
                    placeholder="Tell us about your recipe eg - origin, inspiration"
                    id="description"
                    rows="4"
                    required
                    name="description"
                  />
                </li>

                <li className=" col-lg-6 col-sm-12">
                  <div className="col-lg-11 col-sm-12">
                    <Dropzone
                      onDrop={this.handleDrop}
                      accept="image/jpeg,image/jpg,image/tiff,image/gif,image/png"
                      multiple={false}
                      onDropRejected={this.handleDropRejected}
                      className=" p-10 text-center dropzone bg-light"
                    >
                      Drag a file here or click to upload an image of your food
                    </Dropzone>
                  </div>
                </li>
                <li className="special col-lg-6 col-sm-12">
                  {preview && (
                    <img src={preview} className="col-lg-11 col-sm-12" alt="image preview" />
                  )}
                </li>
                <li className="special col-lg-6 col-sm-12">
                  <label>Category</label>
                  <select name="category" className="col-lg-11 col-sm-12 ">
                    {recipeCategory.map((item, ) => {
                      return (
                        <option value={item} key={item} className="text-capitalize">
                          {item}
                        </option>
                      );
                    })}
                  </select>
                </li>
                <li className=" col-12 ">
                  <input type="submit" value="Submit" id="submit" className="bg-dark btn hovered" />
                </li>
                <li className={`col-12 text-center ${status}`}>
                  <i className="fa fa-spinner fa-pulse fa-3x fa-fw" />
                  <span className="sr-only">Loading...</span>
                </li>
              </ul>
            </form>
          </div>
        </div>
      </section>
    );
  }
}

const mapStateToProps = state => {
  // console.log(state);
  return {
    new_recipe: state.new_recipe,
  };
};
const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({ addRecipe }, dispatch),
});
export default connect(mapStateToProps, mapDispatchToProps)(AddRecipe);
