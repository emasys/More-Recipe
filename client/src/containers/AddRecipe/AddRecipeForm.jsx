import React from 'react';
import Textarea from 'react-textarea-autosize';
import PropTypes from 'prop-types';

import { list as categoryList } from '../../components/CategoryList';

const AddRecipeForm = props => (
  <form onSubmit={props.handleForm}>
    <ul className="form row p-3 mb-10">
      <li className="col-12">
        <label>Recipe Name</label>
        <input
          type="text"
          name="recipe"
          required
          className="col-12"
          id="inputRecipe"
          placeholder="Name of recipe"
        />
        <div className="text-danger" id="recipe_error" />
      </li>
      <li className="col-12">
        <label htmlFor="ingredients" className="col-form-label">
            Ingredients
        </label>
        <Textarea
          placeholder="Add your ingredients and separate with a comma ','"
          className="col-12"
          id="ingredients"
          minRows={1}
          maxRows={10}
          required
          name="ingredients"
        />
        <div className="text-danger" id="ing_error" />
      </li>
      <li className="col-12">
        <label htmlFor="direction">Direction</label>
        <Textarea
          className="col-12"
          placeholder="how to make it happen"
          id="direction"
          minRows={1}
          maxRows={20}
          required
          name="direction"
        />
        <div className="text-danger" id="direction_error" />
      </li>
      <li className="col-12">
        <label htmlFor="description">Description</label>
        <Textarea
          className="col-12"
          placeholder="a short description of your recipe"
          id="description"
          minRows={1}
          maxRows={20}
          required
          name="description"
        />
        <div className="text-danger" id="description_error" />
      </li>
      <li className="special col-12">
        <label>Category</label>
        <select
          name="category"
          className="col-12 "
          style={{ height: '50px' }}
        >
          {categoryList.map(item => (
            <option value={item} key={item} className="text-capitalize">
              {item}
            </option>
          ))}
        </select>
      </li>
      <li className=" col-12 ">
        <input
          type="submit"
          value="Submit"
          id="submit"
          className="bg-dark btn hovered"
        />
      </li>
    </ul>
  </form>
);

AddRecipeForm.propTypes = {
  handleForm: PropTypes.func
};
export default AddRecipeForm;
