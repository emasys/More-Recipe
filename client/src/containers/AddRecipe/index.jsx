import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

//actions
import { addRecipe } from '../../actions/recipeActions';

//component
import Navbar from '../../components/Navbar';
import AddRecipeForm from './AddRecipeForm';
import config from '../../config';

/**
 *
 * @class AddRecipe
 *
 * @extends {Component}
 */
export class AddRecipe extends Component {
  static propTypes = {
    new_recipe: PropTypes.object.isRequired,
    history: PropTypes.object.isRequired,
    addRecipe: PropTypes.func.isRequired,
    auth: PropTypes.object.isRequired
  };

  /**
   * Creates an instance of AddRecipe.
   * @param {any} props
   * @memberof AddRecipe
   */
  constructor(props) {
    super(props);
    this.state = {
      preview: null,
      files: null,
      fileURL: null,
      status: 'fade',
      required: false,
      nameConflict: false,
      buttonDisplay: true
    };
  }

  /**
   * @returns {void}
   *
   *
   * @memberOf AddRecipe
   */
  componentDidMount = () => {
    window.scrollTo(0, 0);
  };

  /**
   * @param {object} nextProps
   *
   * @returns {void}
   *
   * @memberOf AddRecipe
   */
  componentWillReceiveProps = nextProps => {
    if (nextProps.isLoading) {
      this.setState({
        buttonDisplay: false
      });
    } else {
      this.setState({
        buttonDisplay: true
      });
    }
    if (nextProps.new_recipe.new_recipe) {
      if (nextProps.new_recipe.new_recipe.recipe) {
        const { userId } = nextProps.new_recipe.new_recipe.recipe;
        nextProps.history.push(`/profile/${userId}`);
        this.setState({
          nameConflict: false
        });
      }
      if (nextProps.new_recipe.new_recipe.error) {
        this.setState({
          nameConflict: true
        });
      }
    }
  };

  /**
   * clear error messages on field focus
   *
   * @returns {void}
   *
   * @memberOf AddRecipe
   */
  onFocused = () => {
    this.setState({
      nameConflict: false,
      required: false
    });
  };
  /**
   * Handle edit form
   *
   * @param {object} event
   *
   * @returns {void}
   *
   * @memberOf AddRecipe
   */
  handleForm = event => {
    event.preventDefault();
    let data = {
      name: event.target.elements.recipe.value.trim(),
      ingredients: event.target.elements.ingredients.value.trim().toLowerCase(),
      direction: event.target.elements.direction.value.trim(),
      description: event.target.elements.description.value.trim(),
      category: event.target.elements.category.value,
      foodImg: config.DEFAULT_FOOD_IMG
    };
    if (data.name && data.description && data.direction && data.ingredients) {
      this.props.addRecipe(data);
    } else {
      this.setState({
        required: true
      });
    }
  };

  /**
   * @returns {JSX.Element}
   * render react element into the DOM
   * @memberof AddRecipe
   */
  render() {
    return (
      <section className="container ">
        <Navbar className="bg-dark fixed-top" />
        <div
          data-aos="fade-up"
          data-duration="1000"
          className="row p-0 justify-content-center mt-80"
        >
          <div className=" catalog-wrapper col-lg-6 col-md-9 p-0">
            <div className="col-12 text-center AuthInfo">
              <img
                src="https://res.cloudinary.com/emasys/image/upload/v1516439649/mR_2_jwnuce.png"
                alt="logo"
                width="200"
                height="200"
                className="mt-30"
                data-aos="flip-right"
                data-aos-delay="1000"
                data-dos-duration="1000"
              />
              <h3 className="text-white">
                Hey {this.props.auth.authInfo.username}
              </h3>
              <h4 className="mt-10 text-white p-10 ">
                “Cooking is not a science but an art, mistakes are okay, messes
                are fine—the pleasure is in the creating and the sharing of the
                result.” ― Lori Pollan
              </h4>
            </div>
            <AddRecipeForm
              handleForm={this.handleForm}
              onFocus={this.onFocused}
              state={this.state}
            />
          </div>
        </div>
      </section>
    );
  }
}

export const mapStateToProps = state => ({
  new_recipe: state.recipes,
  auth: state.user,
  isLoading: state.isLoading
});

export default connect(mapStateToProps, { addRecipe })(AddRecipe);
