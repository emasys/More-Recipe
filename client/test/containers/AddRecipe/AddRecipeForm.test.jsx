import React from 'react';
import { shallow, mount } from 'enzyme';

// Component
import AddRecipeForm from '../../../src/containers/AddRecipe/AddRecipeForm';
import { AddRecipe } from '../../../src/containers/AddRecipe/index';
import fakeStore from '../../__mocks__/fakeStore';

let wrapper = null;

describe('Test suite for add recipe page', () => {
  const props = {
    handleForm: jest.fn(),
    onFocus: jest.fn(),
    state: { required: false, nameConflict: false }
  };

  beforeEach(() => {
    wrapper = mount(<AddRecipeForm {...props} />);
  });
  it('Should render without errors', () => {
    expect(wrapper.length).toEqual(1);
  });

  it('Should complete form and submit', () => {
    wrapper.find("input[name='recipe']").instance().value = 'How to cook beans';
    wrapper.find("textarea[name='description']").instance().value =
      'some description';
    wrapper.find("textarea[name='direction']").instance().value =
      'some direction';
    wrapper.find("textarea[name='ingredients']").simulate('focus');
    expect(props.onFocus).toHaveBeenCalled();
    wrapper.find("textarea[name='ingredients']").instance().value =
      'water, salt';
    wrapper.find('form').simulate('submit');

    expect(props.handleForm).toHaveBeenCalled();
  });
  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
