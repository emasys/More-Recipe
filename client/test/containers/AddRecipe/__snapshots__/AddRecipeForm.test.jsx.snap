// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test suite for add recipe page should match snapshot 1`] = `
<AddRecipeForm
  handleForm={[Function]}
  onFocus={[Function]}
  state={
    Object {
      "nameConflict": false,
      "required": false,
    }
  }
>
  <form
    onSubmit={[Function]}
  >
    <ul
      className="form row p-10 mb-10"
    >
      <li
        className="col-12"
      >
        <label>
          Recipe Name
        </label>
        <input
          className="col-12"
          id="inputRecipe"
          name="recipe"
          onFocus={[Function]}
          placeholder="Name of recipe"
          required={true}
          type="text"
        />
      </li>
      <li
        className="col-12"
      >
        <label
          htmlFor="description"
        >
          Description
        </label>
        <TextareaAutosize
          className="col-12"
          id="description"
          maxLength={400}
          maxRows={20}
          minRows={1}
          name="description"
          onChange={[Function]}
          onFocus={[Function]}
          onHeightChange={[Function]}
          placeholder="a short description of your recipe (400 characters)"
          required={true}
          useCacheForDOMMeasurements={false}
        >
          <textarea
            className="col-12"
            id="description"
            maxLength={400}
            name="description"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="a short description of your recipe (400 characters)"
            required={true}
            style={
              Object {
                "height": -80,
              }
            }
          />
        </TextareaAutosize>
      </li>
      <li
        className="col-12"
      >
        <label
          className="col-form-label"
          htmlFor="ingredients"
        >
          Ingredients
        </label>
        <TextareaAutosize
          className="col-12"
          id="ingredients"
          maxRows={10}
          minRows={1}
          name="ingredients"
          onChange={[Function]}
          onFocus={[Function]}
          onHeightChange={[Function]}
          placeholder="Add your ingredients and separate with a comma ','"
          required={true}
          useCacheForDOMMeasurements={false}
        >
          <textarea
            className="col-12"
            id="ingredients"
            name="ingredients"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="Add your ingredients and separate with a comma ','"
            required={true}
            style={
              Object {
                "height": -40,
              }
            }
          />
        </TextareaAutosize>
      </li>
      <li
        className="col-12"
      >
        <label
          htmlFor="direction"
        >
          Direction
        </label>
        <TextareaAutosize
          className="col-12"
          id="direction"
          maxRows={20}
          minRows={1}
          name="direction"
          onChange={[Function]}
          onFocus={[Function]}
          onHeightChange={[Function]}
          placeholder="how to make it happen"
          required={true}
          useCacheForDOMMeasurements={false}
        >
          <textarea
            className="col-12"
            id="direction"
            name="direction"
            onChange={[Function]}
            onFocus={[Function]}
            placeholder="how to make it happen"
            required={true}
            style={
              Object {
                "height": -80,
              }
            }
          />
        </TextareaAutosize>
      </li>
      <li
        className="special col-12"
      >
        <label>
          Category
        </label>
        <select
          className="col-12 "
          name="category"
          style={
            Object {
              "height": "50px",
            }
          }
        >
          <option
            className="text-capitalize"
            key="Breakfast"
            value="Breakfast"
          >
            Breakfast
          </option>
          <option
            className="text-capitalize"
            key="Brunch"
            value="Brunch"
          >
            Brunch
          </option>
          <option
            className="text-capitalize"
            key="Lunch"
            value="Lunch"
          >
            Lunch
          </option>
          <option
            className="text-capitalize"
            key="Snacks"
            value="Snacks"
          >
            Snacks
          </option>
          <option
            className="text-capitalize"
            key="Appetisers"
            value="Appetisers"
          >
            Appetisers
          </option>
          <option
            className="text-capitalize"
            key="Dinner"
            value="Dinner"
          >
            Dinner
          </option>
          <option
            className="text-capitalize"
            key="Soups"
            value="Soups"
          >
            Soups
          </option>
          <option
            className="text-capitalize"
            key="Noodles"
            value="Noodles"
          >
            Noodles
          </option>
          <option
            className="text-capitalize"
            key="Rice"
            value="Rice"
          >
            Rice
          </option>
          <option
            className="text-capitalize"
            key="Pasta"
            value="Pasta"
          >
            Pasta
          </option>
          <option
            className="text-capitalize"
            key="Meat"
            value="Meat"
          >
            Meat
          </option>
          <option
            className="text-capitalize"
            key="Poultry"
            value="Poultry"
          >
            Poultry
          </option>
          <option
            className="text-capitalize"
            key="Seafood"
            value="Seafood"
          >
            Seafood
          </option>
          <option
            className="text-capitalize"
            key="Vegetarian"
            value="Vegetarian"
          >
            Vegetarian
          </option>
          <option
            className="text-capitalize"
            key="Sides"
            value="Sides"
          >
            Sides
          </option>
          <option
            className="text-capitalize"
            key="Sauces"
            value="Sauces"
          >
            Sauces
          </option>
          <option
            className="text-capitalize"
            key="Baking"
            value="Baking"
          >
            Baking
          </option>
          <option
            className="text-capitalize"
            key="Desserts"
            value="Desserts"
          >
            Desserts
          </option>
          <option
            className="text-capitalize"
            key="Drinks"
            value="Drinks"
          >
            Drinks
          </option>
          <option
            className="text-capitalize"
            key="Salads"
            value="Salads"
          >
            Salads
          </option>
        </select>
      </li>
      <li
        className=" col-12 text-center "
      >
        <img
          alt="loading..."
          height="50"
          src="https://res.cloudinary.com/emasys/image/upload/v1516647862/Facebook-0.9s-200px_sqqnu9.gif"
          width="50"
        />
      </li>
    </ul>
  </form>
</AddRecipeForm>
`;
